### ISSUE TITLE
15 dynamic lights in screen space limit 

### ISSUE THREAD

**Issue by gabengaGamer** [Facepunch/sbox-issues#497]
**Describe the bug**

If we spawn more than 15 dynamic light sources, we will get a limit-bug.

**To Reproduce**

1. Go to sandbox mode (for convenience)
2. Spawn 15 or more Lamps 
2. Or create in hammer editor 15 and more per pixel lights
3. See bug ?

Video demo: https://www.youtube.com/watch?v=AyPtxJ62BBc

**Screenshots**

![image](https://user-images.githubusercontent.com/54669564/126900128-5798326a-5526-4bc7-be3c-01b65bf3a7d6.png)

**Comment by garrynewman** [comment]
How many do you want to spawn?

**Comment by gabengaGamer** [comment]
More? Complex maps require more dynamic lights.

**Comment by ogniK5377** [comment]
There's a 128 limit to dynamic lights:
![2021-07-23_19-07-4e1b5a34-e28d-44da-8c7d-e6f7fc114aa0-pLRUKzkj](https://user-images.githubusercontent.com/25727384/126900540-16bb8abd-f685-47b2-b193-6ae35442755c.jpg)

**Comment by gabengaGamer** [comment]
they cast no shadows

**Comment by gabengaGamer** [comment]
Turn on the display of shadows through the code and see the limit

**Comment by ogniK5377** [comment]
They're limited based on distance to them

**Comment by gabengaGamer** [comment]
No

**Comment by gabengaGamer** [comment]
Okay, so this will be fixed?

**Comment by kote2345** [comment]
> They're limited based on the distance to them.
> 
>  2021-07-25.23-24-22.mp4 
> The hard coded shader limit for lights is currently at 128 dynamic lights

You can spawn a bunch of lamps next to you and if there are more than 8-10 lamps, then some lamps stop emit light. That is, it does not seem that the problem is in the distance. Here is another video where I fly far enough and only 10 pieces emit light at a time. Same thing happens in Half-Life Alyx, but in hla perhaps reason is different
https://imgur.com/a/D7X0lCA

**Comment by Rohansi** [comment]
What exactly do you need so many dynamic, shadow casting lights for? And why do you absolutely need that many at any given moment? "Complex maps" doesn't answer these.

The only problem I see here so far is they should probably downgrade to shadowless once over the limit instead of doing nothing.

**Comment by gabengaGamer** [comment]
I showed the limit and that's it, situations are different. Let's give an example, there are 10 players on the server, they turned on their flashlights, there are several dynamic sources next to them (6-10), what do we get in the end? limit

**Comment by Rohansi** [comment]
IMO that situation isn't an issue if the lights automatically downgrade to shadowless.  You'll still get light just no extra shadows. Realistically nobody is going to notice this because you have 10 players with flashlights in a very lit area anyway...

There's always going to be a pretty small limit here because more lights will noticeably decrease performance. It's unavoidable without some very significant engine changes. I don't see this changing any time soon unless you can come up with an actual gamemode which is negatively affected by this limit and cannot work any other way.

**Comment by gabengaGamer** [comment]
Light without shadows will look poor

**Comment by gabengaGamer** [comment]
![image](https://user-images.githubusercontent.com/54669564/126913621-e88faf8f-6087-4145-8448-ee90f9dba0ab.png)

there shouldn't have been a limit

**Comment by garrynewman** [comment]
There is a rendering limit. This is for performance reasons, there's always going to be a limit. Whether we can raise it is another question.

The renderer takes the closest dynamic lights and renders those, rather than every light in the map. This is usually fine because lights are normally spread out except for very specific situations. 

I don't know if there's a technical reason the limit is 16, that seems like a decent number of lights to me - but maybe it's one of those alyx limits and we can raise it @samzanemesis.

Either way having an unlimited number of lights rendered isn't something we're going to support, at least not on the forward pipeline.

**Comment by gabengaGamer** [comment]
In that case, I don't see any more sense in this report. I hope this will be fixed in the future.

**Comment by ischmal** [comment]
> What exactly do you need so many dynamic, shadow casting lights for? And why do you absolutely need that many at any given moment? "Complex maps" doesn't answer these.

Does it really matter what someone wants to use it for? We're not idiots. We all know everything has a limit, and we all know how to optimize and make compromises for the best end result.

Having said that, Garry's Mod is considerably more generous with dynamic and projected light limits. And yes, the more you add the more it degrades performance, we are aware. But that really should be on _us_ to implement appropriately. 

The bottom line is this: if someone's hardware supports more graphically intensive operations, then they should be permitted to make the most of it. There is no logical reason not to do it.

**Comment by stopmotio** [comment]
Does anyone think it could be a good idea to add it as a slider to the graphics menu when that arrives? I think it would be a great way to allow scenebuilders to really use as many lamps as needed without random people lagging people with so many lamps. Either way, on the bright side, that's like 7 more lamps than gmod ever did!

**Comment by bellum128** [comment]
> But that really should be on _us_ to implement appropriately

Exactly. This should be a simple convar tied to the render limit with a defualt of say 10. But if I want to set it to 1000, I should be able to do that even if it gives me 1 frame per minute.

**Comment by stopmotio** [comment]
> > Yea I wish. Maybe with the point lights but those also were not nearly as good as S&box. `env_projectedtexture` had a limit of 9, versus the S&box equivalent of 16.
> 
> Is this the case? IIRC the limit was removed or significantly raised around 2018-2019. I remember because that's when I was able to do this:
> 
> ![VgJNiyyK01](https://user-images.githubusercontent.com/5350086/136499701-cdb3b4e2-1d10-466b-814a-4cddb2d8182d.gif)
> 
> Maybe it was scaled back later on, but I know I was able to spawn dozens of projected textures without much of a problem.

That could be within the light limit. If it's that recent it's possible, but according to the VDC that's the limit.
